codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/JsonShapeSerVisitor.java

```diff
@@ -0,0 +1,198 @@
+        writer.write("");
+
+        writer.openBlock("for key := range v {", "}", () -> {
+            writer.write("om := object.Key(key)");
+
+            // Null values in maps should be serialized as such. Enums can't be null, so we don't bother
+            // putting this in for their case.
+            if (!target.hasTrait(EnumTrait.class)) {
+                writer.openBlock("if vv := v[key]; vv == nil {", "}", () -> {
+                    writer.write("om.Null()");
+                    writer.write("continue");
+                });
+            }
+
+            target.accept(getMemberSerVisitor(shape.getValue(), "v[key]", "om"));
+        });
+
+        writer.write("return nil");
+    }
+
+    @Override
+    protected void serializeStructure(GenerationContext context, StructureShape shape) {
+        GoWriter writer = context.getWriter();
+
+        writer.write("object := value.Object()");
+        writer.write("defer object.Close()");
+        writer.write("");
+
+        // Use a TreeSet to sort the members.
+        Set<MemberShape> members = new TreeSet<>(shape.getAllMembers().values());
+        for (MemberShape member : members) {
+            Shape target = context.getModel().expectShape(member.getTarget());
+            String serializedMemberName = getSerializedMemberName(member);
+            writeSafeMemberAccessor(context, member, "v", (operand) -> {
+                writer.write("ok := object.Key($S)", serializedMemberName);
+                target.accept(getMemberSerVisitor(member, operand, "ok"));
+            });
+            writer.write("");
+        }
+
+        writer.write("return nil");
+    }
+
+    private String getSerializedMemberName(MemberShape memberShape) {
+        Optional<JsonNameTrait> jsonNameTrait = memberShape.getTrait(JsonNameTrait.class);
+        return jsonNameTrait.isPresent() ? jsonNameTrait.get().getValue() : memberShape.getMemberName();
+    }
+
+    @Override
+    protected void serializeUnion(GenerationContext context, UnionShape shape) {
+        GoWriter writer = context.getWriter();
+        SymbolProvider symbolProvider = context.getSymbolProvider();
+        Symbol symbol = symbolProvider.toSymbol(shape);
+        writer.addUseImports(SmithyGoDependency.FMT);
+
+        writer.write("object := value.Object()");
+        writer.write("defer object.Close()");
+        writer.write("");
+
+        writer.openBlock("switch uv := v.(type) {", "}", () -> {
+            // Use a TreeSet to sort the members.
+            Set<MemberShape> members = new TreeSet<>(shape.getAllMembers().values());
+            for (MemberShape member : members) {
+                Shape target = context.getModel().expectShape(member.getTarget());
+                Symbol memberSymbol = symbolProvider.toSymbol(member);
+                String exportedMemberName = symbol.getName() + symbolProvider.toMemberName(member);
+
+                writer.openBlock("case *$L:", "", exportedMemberName, () -> {
```

Comment:
`exportedMemberName` is the name of type for the union member correct?

Reply:
Yes